libgraphics/include/pivot/graphics/VulkanException.hxx

```diff
@@ -1,20 +1,53 @@
 #pragma once
 
 #include <stdexcept>
-#include <vulkan/vulkan.hpp>
 
-namespace pivot::graphics
+namespace pivot
 {
+/// Base class for pivot exceptions
+class PivotException : public std::exception
+{
+public:
+    PivotException() = delete;
+    /// dtor
+    virtual ~PivotException(){};
+    /// main ctor
+    PivotException(const std::string &msg): msg(msg) {}
+    /// return the scope of the exception
+    virtual const std::string_view getScope() const noexcept = 0;
+    /// return the kind of the exception
+    virtual const std::string_view getKind() const noexcept = 0;
+    /// return both of the above
+
+    /// return the error message
+    const char *what() const noexcept override { return msg.c_str(); }
+
+private:
+    const std::string msg;
+};
 
-/// @class VulkanException
-/// @brief A custom exception throw when a error occurred in the graphical engine
-///
-/// @see OutOfDateSwapchainError
-/// @see TooManyObjectInSceneError
-struct VulkanException : public std::runtime_error {
-    using std::runtime_error::runtime_error;
-    /// Constructor converting a vk::Result into a string and use it as error message
-    VulkanException(const vk::Result &er): VulkanException(vk::to_string(er)){};
+/// Pivot Logic Error
+class LogicError : public PivotException
+{
+    using PivotException::PivotException;
+    const std::string_view getKind() const noexcept final { return "Logic"; };
 };
 
-}    // namespace pivot::graphics
+/// Pivot Runtime Error
+class RuntimeError : public PivotException
+{
+    using PivotException::PivotException;
+    const std::string_view getKind() const noexcept final { return "Logic"; };
+};
+
+}    // namespace pivot
+
+#define PIVOT_EXCEPTION(kind, name)                                                 \
+    struct name##Error : public ::pivot::kind {                                     \
+        using kind::kind;                                                           \
+        const std::string_view getScope() const noexcept override { return #name; } \
+    };
```

Comment:
I'm not sure if I like macros defining types with names not passed as parameter.